csds-client/client/client.go

```diff
@@ -0,0 +1,255 @@
+		requestFile: *requestFilePtr,
+		requestYaml: *requestYamlPtr,
+		jwt:         *jwtPtr,
+		configFile:  *configFilePtr,
+		monitorFreq: *monitorFreqPtr,
+	}
+
+	return f
+}
+
+// parseNodeMatcher parses the csds request yaml to nodematcher
+func (c *Client) parseNodeMatcher() error {
+	if c.info.requestFile == "" && c.info.requestYaml == "" {
+		return fmt.Errorf("missing request yaml")
+	}
+
+	var nodematchers []*envoy_type_matcher.NodeMatcher
+	if err := parseYaml(c.info.requestFile, c.info.requestYaml, &nodematchers); err != nil {
+		return err
+	}
+
+	c.nm = nodematchers
+
+	// check if required fields exist in nodematcher
+	switch c.info.platform {
+	case "gcp":
+		switch c.info.uri {
+		case "trafficdirector.googleapis.com:443":
+			keys := []string{"TRAFFICDIRECTOR_GCP_PROJECT_NUMBER", "TRAFFICDIRECTOR_NETWORK_NAME"}
+			for _, key := range keys {
+				if value := getValueByKeyFromNodeMatcher(c.nm, key); value == "" {
+					return fmt.Errorf("Missing field %v in NodeMatcher", key)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// connWithAuth connects to uri with authentication
+func (c *Client) connWithAuth() error {
+	var scope string
+	switch c.info.authnMode {
+	case "jwt":
+		if c.info.jwt == "" {
+			return fmt.Errorf("missing jwt file")
+		}
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewServiceAccountFromFile(c.info.jwt, scope)
+			if err != nil {
+				return err
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return nil
+		}
+	case "auto":
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewApplicationDefault(context.Background(), scope) // Application Default Credentials (ADC)
+			if err != nil {
+				return err
+			}
+
+			// parse GCP project number as header for authentication
+			var key string
+			switch c.info.uri {
+			case "trafficdirector.googleapis.com:443":
+				key = "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER"
+			}
+			if projectNum := getValueByKeyFromNodeMatcher(c.nm, key); projectNum != "" {
+				c.md = metadata.Pairs("x-goog-user-project", projectNum)
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return fmt.Errorf("Auto authentication mode for this platform is not supported. Please use jwt_file instead")
+		}
+	default:
+		return fmt.Errorf("Invalid authn_mode")
+	}
+}
+
+// New creates a new client
+func New() (*Client, error) {
+	c := &Client{
+		info: ParseFlags(),
```

Comment:
It is better to parse flags in the main, and pass it to the new as struct with parameters.

If this module would need to be reused - then it is tightly coupled with the chosen flags. While flags should be a choice of a binary, not a module. By decoupling, you'll let binary define its own set of flags and decide which options of the module user will want to populate.

This will not change functionality now, but will make code more reusable in the future.

Reply:
Sorry, I'm a little confusing. If I move the `ParseFlags()` to main and also define the type `Flag` in main, how can I use this struct in client properly? Since I need to define `New()` as `func New(flags Flag) (*Client, error)` right? If I only pass the flags to `New()` as struct like `func New(flags struct{}) (*Client, error)` but not a new type of Flag, I cannot access the fields in struct by `c.Info.uri` since I didn't define the fields in this struct in client.
Sorry for that if I misunderstood something, this is my first time to learn and use Go, there're still a lot of things for me to learn.

Reply:
Cool. Make sense. Thanks for your explanation in detail! 
I think I can implement this in next PR, since this PR is already pretty big in size. Maybe just want it to be more basic to make it easier to review.
There are already some feature enhancement planned to be included in next PR. I think I can also include this to define better boundaries and make it more reusable.

Reply:
Sounds good, let's refactor in other PR